src/pybind/mgr/cephadm/module.py

```diff
@@ -1682,24 +1659,38 @@ def add_host(self, spec):
             raise OrchestratorError('New host %s (%s) failed check: %s' % (
                 spec.hostname, spec.addr, err))
 
-        self.inventory[spec.hostname] = spec.to_json()
-        self._save_inventory()
+        self.inventory.add_host(spec)
         self.cache.prime_empty_host(spec.hostname)
         self.offline_hosts_remove(spec.hostname)
         self.event.set()  # refresh stray health check
         self.log.info('Added host %s' % spec.hostname)
         return "Added host '{}'".format(spec.hostname)
 
-    @async_completion
-    def remove_host(self, host):
-        # type: (str) -> str
-        """
-        Remove a host from orchestrator management.
+    def _verify_host_is_empty(self, host):
+        daemons = dict(self.cache.get_daemons_with_volatile_status())
+        daemons_on_host = daemons.get(host)
+        if daemons_on_host:
+            names = ', '.join(daemons_on_host.keys())
+            s = f'Daemons still on host `{host}`: {names}\n'
+            s += 'Please remove all daemons from this host, before removing the host.\n'
+            s += '(Add --force to ignore this message)'
+            raise OrchestratorValidationError(s)
+
+        specs = self.spec_store.specs.values()
+        placements = [(s.service_name(), s.placement) for s in specs if s.placement is not None]
+        for name, placement in placements:
+            if any([host == h.hostname for h in placement.hosts]):
+                s = f'Service {name} still specifies its placement to be on host `{host}`\n'
+                s += 'Please alter the placement specification, before removing the host.\n'
+                s += '(Add --force to ignore this message)'
```

Comment:
I see why we want to have a way to skip this, but here again, this may lead into chaos..

Reply:
maybe hint to `ceph orch host drain ` would be helpful?